### PR TITLE
Add support for Kopia location profile

### PR DESCRIPTION
**Pull request by PrasadG193**
## Change Overview

- Add new Location and Credential type - kopia 
- Add a way to pass connection info and creds to connect to Kopia server through Profile for kando operations

Depends on: https://github.com/kanisterio/kanister/pull/944

## Pull request type

Please check the type of change your PR introduces:
- [x] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [x] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Issues

- #XXX

## Test Plan

- [x] :muscle: Manual
- [ ] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

File: pkg/apis/cr/v1alpha1/types.go [hunk 1]
```diff
@@ -257,6 +258,7 @@ type Location struct {
 	Endpoint string       `json:"endpoint"`
 	Prefix   string       `json:"prefix"`
 	Region   string       `json:"region"`
+	Hostname string       `json:"hostname"`
```
Comment (pavannd1) [inline pkg/apis/cr/v1alpha1/types.go:261]: Could this go into the credential as well? That way, the location could stay the same

File: pkg/apis/cr/v1alpha1/types.go [hunk 2]
```diff
@@ -265,13 +267,15 @@ type CredentialType string
...
 
 // Credential
 type Credential struct {
-	Type    CredentialType   `json:"type"`
-	KeyPair *KeyPair         `json:"keyPair,omitempty"`
-	Secret  *ObjectReference `json:"secret,omitempty"`
+	Type         CredentialType   `json:"type"`
+	KeyPair      *KeyPair         `json:"keyPair,omitempty"`
+	Secret       *ObjectReference `json:"secret,omitempty"`
+	KopiaSecrets *KopiaSecret     `json:"kopiaSecrets,omitempty"`
```
Comment (pavannd1) [inline pkg/apis/cr/v1alpha1/types.go:278]:
```suggestion
	KopiaServerSecret *KopiaServerSecret     `json:"kopiaSecrets,omitempty"`
```
Can we make this and the `param.Profile` consistent?

File: pkg/param/param.go [hunk 3]
```diff
@@ -108,13 +108,15 @@ type CredentialType string
...
 
 // Credential resolves the storage
 type Credential struct {
-	Type    CredentialType
-	KeyPair *KeyPair
-	Secret  *v1.Secret
+	Type        CredentialType
+	KeyPair     *KeyPair
+	Secret      *v1.Secret
+	KopiaServer *StoreServerCreds
```
Comment (pavannd1) [inline pkg/param/param.go:119]: As mentioned above, can we make the names consistent like other fields?

File: pkg/apis/cr/v1alpha1/types.go [hunk 4]
```diff
@@ -265,13 +266,15 @@ type CredentialType string
...
 
 // Credential
 type Credential struct {
-	Type    CredentialType   `json:"type"`
-	KeyPair *KeyPair         `json:"keyPair,omitempty"`
-	Secret  *ObjectReference `json:"secret,omitempty"`
+	Type              CredentialType     `json:"type"`
+	KeyPair           *KeyPair           `json:"keyPair,omitempty"`
+	Secret            *ObjectReference   `json:"secret,omitempty"`
+	KopiaServerSecret *KopiaServerSecret `json:"kopiaSecrets,omitempty"`
```
Comment (pavannd1) [inline pkg/apis/cr/v1alpha1/types.go:277]:
```suggestion
	KopiaServerSecret *KopiaServerSecret `json:"kopiaServerSecret,omitempty"`
```

File: pkg/apis/cr/v1alpha1/types.go [hunk 5]
```diff
@@ -281,6 +284,18 @@ type KeyPair struct {
 	Secret      ObjectReference `json:"secret"`
 }
 
+type KopiaServerSecret struct {
+	Username       string                `json:"username,omitempty"`
+	Hostname       string                `json:"hostname,omitempty"`
+	UserPassPhrase *KopiaServerSecretRef `json:"userPassPhrase,omitempty"`
```
Comment (pavannd1) [inline pkg/apis/cr/v1alpha1/types.go:290]:
```suggestion
	UserPassphrase *KopiaServerSecretRef `json:"userPassphrase,omitempty"`
```

File: pkg/param/param.go [hunk 6]
```diff
@@ -123,6 +125,14 @@ type KeyPair struct {
 	Secret string
 }
 
+// StoreServerCreds contains creds to communicate with Kopia server
+type StoreServerCreds struct {
```
Comment (pavannd1) [inline pkg/param/param.go:129]: Should this be `KopiaServerCreds`?

File: pkg/apis/cr/v1alpha1/types.go [hunk 7]
```diff
@@ -281,6 +284,18 @@ type KeyPair struct {
 	Secret      ObjectReference `json:"secret"`
 }
 
+type KopiaServerSecret struct {
```
Comment (pavannd1) [inline pkg/apis/cr/v1alpha1/types.go:287]: Add comments here and below

File: pkg/param/param.go [hunk 8]
```diff
@@ -451,6 +463,49 @@ func fetchPVCParams(ctx context.Context, cli kubernetes.Interface, namespace, na
 	}, nil
 }
 
+func fetchKopiaCredential(ctx context.Context, cli kubernetes.Interface, ks *crv1alpha1.KopiaServerSecret) (*Credential, error) {
+	if ks == nil {
+		return nil, errors.New("Kopia Secret reference cannot be nil")
+	}
+
+	if ks.UserPassPhrase.Secret == nil {
+		return nil, errors.New("Kopia UserPassPhrase Secret reference cannot be nil")
```
Comment (pavannd1) [inline pkg/param/param.go:472]:
```suggestion
	if ks.UserPassphrase.Secret == nil {
		return nil, errors.New("Kopia UserPassphrase secret reference cannot be nil")
```

File: pkg/param/param.go [hunk 9]
```diff
@@ -451,6 +463,49 @@ func fetchPVCParams(ctx context.Context, cli kubernetes.Interface, namespace, na
...
+	}
+
+	passSecret, err := cli.CoreV1().Secrets(ks.UserPassPhrase.Secret.Namespace).Get(ctx, ks.UserPassPhrase.Secret.Name, metav1.GetOptions{})
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to fetch the secret %s/%s", ks.UserPassPhrase.Secret.Namespace, ks.UserPassPhrase.Secret.Name)
```
Comment (pavannd1) [inline pkg/param/param.go:477]:
```suggestion
	passSecret, err := cli.CoreV1().Secrets(ks.UserPassphrase.Secret.Namespace).Get(ctx, ks.UserPassphrase.Secret.Name, metav1.GetOptions{})
	if err != nil {
		return nil, errors.Wrapf(err, "Failed to fetch the secret %s/%s", ks.UserPassphrase.Secret.Namespace, ks.UserPassphrase.Secret.Name)
```

File: pkg/param/param.go [hunk 10]
```diff
@@ -451,6 +463,49 @@ func fetchPVCParams(ctx context.Context, cli kubernetes.Interface, namespace, na
...
+	}
+	//if err = secrets.ValidateCredentials(s); err != nil {
+	//	return nil, err
+	//}
```
Comment (pavannd1) [inline pkg/param/param.go:481]:
```suggestion

```

File: pkg/param/param.go [hunk 11]
```diff
@@ -451,6 +463,49 @@ func fetchPVCParams(ctx context.Context, cli kubernetes.Interface, namespace, na
...
+	password, ok := passSecret.Data[ks.UserPassPhrase.Key]
+	if !ok {
+		return nil, errors.New("Failed to fetch userPassPhrase from secret")
```
Comment (pavannd1) [inline pkg/param/param.go:484]:
```suggestion
		return nil, errors.New("Failed to fetch user passphrase from secret")
```

File: pkg/param/param.go [hunk 12]
```diff
@@ -451,6 +463,49 @@ func fetchPVCParams(ctx context.Context, cli kubernetes.Interface, namespace, na
...
+	}
+
+	if ks.TLSCert == nil || ks.TLSCert.Secret == nil {
+		return nil, errors.New("Kopia TLS cert Secret reference cannot be nil")
```
Comment (pavannd1) [inline pkg/param/param.go:488]:
```suggestion
		return nil, errors.New("Kopia TLS cert secret reference cannot be nil")
```

**Review by pavannd1 — COMMENTED** [review]
Some nits

File: pkg/apis/cr/v1alpha1/types.go (hunk 4 above)
Comment (PrasadG193) [reply pkg/apis/cr/v1alpha1/types.go:277]: :man_facepalming: Sorry about this.

File: pkg/kando/location_pull.go [hunk 13]
```diff
@@ -59,7 +61,14 @@ func runLocationPull(cmd *cobra.Command, args []string) error {
 		return err
 	}
 	s := pathFlag(cmd)
+	id := backupIDFlag(cmd)
 	ctx := context.Background()
+	if p.Location.Type == crv1alpha1.LocationTypeKopia {
```
Comment (pavannd1) [inline pkg/kando/location_pull.go:66]: Can we merge the profile changes with this PR and open a separate one for these changes?

File: pkg/kando/location_push.go [hunk 14]
```diff
@@ -52,6 +53,12 @@ func runLocationPush(cmd *cobra.Command, args []string) error {
 	}
 	s := pathFlag(cmd)
 	ctx := context.Background()
+	if p.Location.Type == crv1alpha1.LocationTypeKopia {
```
Comment (pavannd1) [inline pkg/kando/location_push.go:56]: Same as above, could be a separate PR

File: pkg/kando/location_pull.go (hunk 13 above)
Comment (PrasadG193) [reply pkg/kando/location_pull.go:66]: Sounds good!

File: pkg/kando/location_pull.go [hunk 15]
```diff
@@ -79,3 +80,14 @@ func locationPull(ctx context.Context, p *param.Profile, path string, target io.
 func kopiaLocationPull(ctx context.Context, backupID, path string, target io.Writer) error {
 	return kopia.Read(ctx, backupID, path, target)
 }
+
+// connectToKopiaServer connects to the kopia server with given creds
+// nolint:unused,deadcode
+func connectToKopiaServer(ctx context.Context, kp *param.Profile) error {
+	log.Debug().Print("Connecting to kopia server")
```
Comment (pavannd1) [inline pkg/kando/location_pull.go:87]: This may cause issues during `locationPull` if we set the log level to debug. Since the data is coming through `stdout` and we are trying to print to `stdout`, the DB data might get affected.
We should remove the log to make it work.

File: pkg/kando/location_pull.go [hunk 16]
```diff
@@ -79,3 +80,14 @@ func locationPull(ctx context.Context, p *param.Profile, path string, target io.
...
+	err := kopia.ConnectToAPIServer(ctx, kp.Credential.KopiaServerSecret.Cert, kp.Credential.KopiaServerSecret.Password, kp.Credential.KopiaServerSecret.Hostname, kp.Location.Endpoint, kp.Credential.KopiaServerSecret.Username)
+	if err == nil {
+		log.Debug().Print("Connected to kopia server")
```
Comment (pavannd1) [inline pkg/kando/location_pull.go:90]: Same issue as above.

**Review by pavannd1 — COMMENTED** [review]
LGTM 👍🏼

File: pkg/kando/location_pull.go (hunk 15 above)
Comment (PrasadG193) [reply pkg/kando/location_pull.go:87]: Oh ok, will remove that

**Review by pavannd1 — APPROVED** [review]
